editor/test/integration/lsp_test.clj

```diff
@@ -0,0 +1,363 @@
+                             (let [message (cond-> {:jsonrpc "2.0"
+                                                    :method method}
+                                                   params
+                                                   (assoc :params params))]
+                               (a/put! base-sink (json/write-str message))))]
+        (lsp.jsonrpc/make (fn get-request-method [method]
+                            (when-let [f (request-handlers method)]
+                              (fn respond [params]
+                                (f params notify-client!))))
+                          base-source
+                          base-sink)
+        (reify lsp.server/Connection
+          (input-stream [_] public-in)
+          (output-stream [_] public-out)
+          (dispose [_]))))))
+
+(def default-handlers
+  {"initialize" (fn [_ _] {:capabilities {:textDocumentSync 2}})
+   "textDocument/didOpen" (fn [{{:keys [uri]} :textDocument} notify!]
+                            (notify! "textDocument/publishDiagnostics"
+                                     {:uri uri
+                                      :diagnostics [{:range {:start {:line 0 :character 0}
+                                                             :end {:line 0 :character 1}}
+                                                     :message "It's a bad start!"
+                                                     :severity 1}]}))
+   "initialized" (constantly nil)
+   "shutdown" (constantly nil)
+   "exit" (constantly nil)})
+
+(deftest lsp-server-test
+  (testing "Initialize + open text document -> should publish diagnostics"
+    (tu/with-loaded-project "test/resources/lsp_project"
+      (let [in (a/chan 10)
+            out (a/chan 10)]
+        (lsp.server/make
+          project
+          (make-test-server-launcher default-handlers)
+          in out
+          :on-publish-diagnostics #(apply vector :on-publish-diagnostics %&)
+          :on-initialized #(vector :on-initialized %))
+        (is (= [[:on-initialized
+                 {:text-document-sync {:open-close true
+                                       :change :incremental}}]
+                [:on-publish-diagnostics
+                 (tu/resource workspace "/foo.json")
+                 [(assoc (data/->CursorRange (data/->Cursor 0 0) (data/->Cursor 0 1))
+                    :type :diagnostic :hoverable true :message "It's a bad start!" :severity :error)]]]
+               (async-support/eventually
+                 (a/go
+                   (>! in (lsp.server/open-text-document (tu/resource workspace "/foo.json") ["{\"a\": 1}"]))
+                   (<! (a/timeout 10))
+                   (a/close! in)
+                   (<! (a/reduce conj [] out))))))))))
+
+(g/defnode LSPViewNode
+  (property diagnostics g/Any (default [])))
+
+(deftest start-open-order-test
+  (tu/with-loaded-project "test/resources/lsp_project"
+    (with-redefs [ui/do-run-later (bound-fn [f] (f))]
```

Comment:
We also have `test-util/with-ui-run-later-rebound` and `test-util/run-event-loop!` that might help with this sort of thing. In those, run-later will enqueue stuff that will actually run later, perhaps providing an order of events more similar to the real thing.